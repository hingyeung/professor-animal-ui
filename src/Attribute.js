import React, { Component } from 'react';

class Attribute extends Component {
  constructor(props) {
    super(props);

    this.name = props.name;
  }

  render() {
    return (
      <div class="attribute-row">
        <div class="attribute-cell">
          {this.name}
        </div>
        <div class="attribute-cell">
          <label htmlFor={this.name + '-yes'}>YES</label>
          <input type="radio" name={this.name} value="yes" id={this.name + '-yes'}/>
        </div>
        <div class="attribute-cell">
          <label htmlFor={this.name + '-no'}>NO</label>
          <input type="radio" name={this.name} value="no" id={this.name + '-no'}/>
        </div>
      </div>
    )
  }
}

export default Attribute;