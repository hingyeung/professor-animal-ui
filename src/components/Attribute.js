import React, {Component} from 'react';

class Attribute extends Component {
  constructor(props) {
    super(props);

    this.props = props;
    this.attribute = props.attribute;
    this.type = props.type;
    this.onAttributeChange = props.onAttributeChange;
  }

  onChange(attrType, attrName, e) {
    this.onAttributeChange(attrType, attrName, e.target.value);
  }

  render() {
    const yesId = this.type + '-' + this.attribute.name + '-yes',
          noId = this.type + '-' + this.attribute.name + '-no';

    return (
      <tr className="attribute-row row">
        <td className="attribute-cell col-sm-6">
          { this.attribute.name }
        </td>
        <td className="attribute-cell col-sm-3">
          <input className="form-check-label" type="radio" name={ this.type + '-' + this.attribute.name } value="yes"
                 id={ yesId }
                 onChange={ (e) => this.onChange(this.type, this.attribute.name, e) }
                 defaultChecked={this.attribute.value}/>
          <label className="form-check-label" htmlFor={ yesId }>YES</label>
        </td>
        <td className="attribute-cell col-sm-3">
          <input className="form-check-label" type="radio" name={ this.type + '-' + this.attribute.name } value="no"
                 id={ noId } onChange={ (e) => this.onChange(this.type, this.attribute.name, e) }
                 defaultChecked={!this.attribute.value}/>
          <label className="form-check-label" htmlFor={ noId }>NO</label>
        </td>
      </tr>
    )
  }
}

export default Attribute;