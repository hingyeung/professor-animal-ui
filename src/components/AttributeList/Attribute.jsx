import React from 'react';

function Attribute(props) {
  const attribute = props.attribute;
  const type = props.type;
  const onAttributeChange = props.onAttributeChange;
  const yesId = props.type + '-' + props.attribute.name + '-yes',
    noId = props.type + '-' + props.attribute.name + '-no',
    // Initially this.state.attributeValue is undefined (before animals are loaded),
    // and it makes the radio button "uncontrolled", then later when this.state.attributeValue
    // is populated with either true or false and the radio button becomes "controlled".
    // ReactJS doesn't like it and would give warning. Making sure its value is either true
    // or false prevents the problem.
    // https://github.com/facebook/react/issues/6779#issuecomment-222162404
    yesIsChecked = props.attribute.value ? props.attribute.value : false;

  function onChange(e) {
    onAttributeChange(type, attribute.name, e.target.value);
  }

  return (
    <tr className="attribute-row row">
      <td className="attribute-cell col-sm-6">
        { attribute.name }
      </td>
      <td className="attribute-cell col-sm-3">
        <input className="form-check-label" type="radio" name={ type + '-' + attribute.name } value="yes"
               id={ yesId }
               onChange={ onChange }
               checked={ yesIsChecked }/>
        <label className="form-check-label" htmlFor={ yesId }>YES</label>
      </td>
      <td className="attribute-cell col-sm-3">
        <input className="form-check-label" type="radio" name={ type + '-' + attribute.name } value="no"
               id={ noId } onChange={ onChange }
               checked={ !yesIsChecked }/>
        <label className="form-check-label" htmlFor={ noId }>NO</label>
      </td>
    </tr>
  )
}

export default Attribute;