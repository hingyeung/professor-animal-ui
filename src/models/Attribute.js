class Attribute {
  constructor(name, defaultValue) {
    this._name = name;
    this._defaultValue = defaultValue;
    this._value = '';
  }

  get name() {
    return this._name;
  }

  get defaultValue() {
    return this._defaultValue;
  };

  get value() {
    return this._value;
  }

  set value(value) {
    this._value = value;
  }
}

export default Attribute;