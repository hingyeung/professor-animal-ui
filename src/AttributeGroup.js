import React, { Component } from 'react';
import Attribute from './Attribute';

class AttributeGroup extends Component {
  constructor(props) {
    super(props);
    this.type = props.type;
    this.attributes = props.attributes;
  }

  render() {
    return this.attributes.map((attribute, index) => {
      return <Attribute key={index} name={attribute.name}/>
    })
  }
}

export default AttributeGroup;