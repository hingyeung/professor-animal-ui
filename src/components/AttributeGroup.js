import React, {Component} from 'react';
import Attribute from './Attribute';

class AttributeGroup extends Component {
  constructor(props) {
    super(props);
    this.type = props.type;
    this.attributes = props.attributes;
    this.onAttributeChange = props.onAttributeChange;
  }

  render() {
    const attributesContent = this.attributes.map((attribute, index) => {
      return <Attribute key={ index } type={ this.type } name={ attribute.name }
                        onAttributeChange={ this.onAttributeChange }/>
    });

    // make a table with header here
    return (
      <div>
        <div>
          <div>
          { this.type }
          </div>
        </div>
        { attributesContent }
      </div>
    )
  }
}

export default AttributeGroup;