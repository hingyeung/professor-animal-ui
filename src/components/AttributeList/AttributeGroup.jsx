import React from 'react';
import Attribute from './Attribute';
import AttributeModel from 'models/Attribute';
import NewAttributeValue from './NewAttributeValue';

function AttributeGroup(props) {
  const attributeType = props.attributeType;
  const attributes = props.attributes;
  const onAttributeChange = props.onAttributeChange;
  const onAddNewAttributeFromParent = props.onAddNewAttribute;

  function _onAddNewAttribute(attribute) {
    onAddNewAttributeFromParent(attributeType, attribute);
  }

  const attributesContent = Object.keys(attributes).map((attributeName, index) => {
    return <Attribute key={ index } type={ attributeType }
                      attribute={ new AttributeModel(attributeName, attributes[attributeName]) }
                      animalId={ props.animalId }
                      onAttributeChange={ props.onAttributeChange }/>
  });

  // make a table with header here
  return (
    <table className="table table-striped">
      <thead>
      <tr className="row">
        <th colSpan="3" className="col-sm-12">{ attributeType }</th>
      </tr>
      </thead>
      <tbody>
      { attributesContent }
      <NewAttributeValue newAttributeIndex={ Object.keys(attributes).length }
                         onAdd={ _onAddNewAttribute }/>
      </tbody>
    </table>
  )
}

export default AttributeGroup;