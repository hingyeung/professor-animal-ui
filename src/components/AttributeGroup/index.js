import AttributeGroup from './AttributeGroup';

export default AttributeGroup;