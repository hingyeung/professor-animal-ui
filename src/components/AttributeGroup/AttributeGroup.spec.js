import React from 'react';
import {shallow, mount} from 'enzyme';
import AttributeGroup from "./AttributeGroup";
import Attribute from "../../models/Attribute";

const mockOnAttributeChange = jest.fn();
const mockOnNewAttributeAdded = jest.fn();
const attributes = {a: true, b: false, c: true, d: false};

jest.mock('components/NewAttributeValue');

describe('AttributeGroup', function () {
  it('should render properly', function () {
    const wrapper = shallow(
      <AttributeGroup attributeType={ 'attributeType' } attributes={ attributes }
                      onAttributeChange={ mockOnAttributeChange } onNewAttributeAdded={ mockOnNewAttributeAdded }/>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onAddNewAttributeFromParent in props when new attribute is added', function () {
    const wrapper = mount(
      <AttributeGroup attributeType={ 'attributeType' } attributes={ attributes }
                      onAttributeChange={ mockOnAttributeChange } onNewAttributeAdded={ mockOnNewAttributeAdded }/>
    );

    // the stub version of NewAttributeValue is a button that call its onNewAttributeAdded() when clicked
    wrapper.find('NewAttributeValue #mock-new-attribute-value').simulate('click');
    expect(mockOnNewAttributeAdded).toHaveBeenCalledWith('attributeType', new Attribute('new attribute', true));
  })
});