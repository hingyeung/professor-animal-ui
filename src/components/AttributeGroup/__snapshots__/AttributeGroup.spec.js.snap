// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeGroup should render properly 1`] = `
<table
  className="table table-striped"
>
  <thead>
    <tr
      className="row"
    >
      <th
        className="attribute-type col-sm-12"
        colSpan="3"
      >
        attributeType
      </th>
    </tr>
  </thead>
  <tbody>
    <Attribute
      attribute={
        Attribute {
          "_name": "a",
          "_value": true,
        }
      }
      key="0"
      onAttributeChange={[Function]}
      type="attributeType"
    />
    <Attribute
      attribute={
        Attribute {
          "_name": "b",
          "_value": false,
        }
      }
      key="1"
      onAttributeChange={[Function]}
      type="attributeType"
    />
    <Attribute
      attribute={
        Attribute {
          "_name": "c",
          "_value": true,
        }
      }
      key="2"
      onAttributeChange={[Function]}
      type="attributeType"
    />
    <Attribute
      attribute={
        Attribute {
          "_name": "d",
          "_value": false,
        }
      }
      key="3"
      onAttributeChange={[Function]}
      type="attributeType"
    />
    <NewAttributeValue
      attributeType="attributeType"
      newAttributeIndex={4}
      onNewAttributeAdded={[Function]}
    />
  </tbody>
</table>
`;
